staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/managedfieldsupdater.go

```diff
@@ -72,9 +72,16 @@ func (f *managedFieldsUpdater) Apply(liveObj, appliedObj runtime.Object, managed
 	if err != nil {
 		return object, managed, err
 	}
+
+	_, managerInFields := managed.Fields()[fieldManager]
+	_, managerInTimes := managed.Times()[fieldManager]
 	if object != nil {
 		managed.Times()[fieldManager] = &metav1.Time{Time: time.Now().UTC()}
 	} else {
+		if managerInFields && !managerInTimes {
```

Comment:
I'm not very familiar with this code, but it looks strange to put this logic in the branch that is calling `RemoveObjectManagedFields` ... are we sure this is the branch where a new manager is *adding* co-ownership of some fields?

this definitely needs really good unit tests around the scenario being exercised

Reply:
Thanks for the prompt review!

I have the same feeling that the two branches are a little confusing to follow. Let me try my best to clarify.
- Since `managedFieldsUpdater` implements the `Manager` interface, its method `(f *managedFieldsUpdater) Apply` should return ‘the new object with managedFields removed, and the object's new proposed managedFields **separately**’:
   https://github.com/kubernetes/kubernetes/blob/7ee17ce9b7c2a22e63e2bbd79d48d3fe349a9386/staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/manager.go#L45-L51
- The `else` branch says: Since the live object is not touched (signaled by `object == nil`, detailed below), the merged object should be the same as the live object. So first make a deep copy of `liveObj`, then remove its managedFields to fit the `Manager` interface.

Details on `object == nil`:
These lines say that `nil` signals the situation that a merged object and the current live object being the same.
https://github.com/kubernetes/kubernetes/blob/7ee17ce9b7c2a22e63e2bbd79d48d3fe349a9386/vendor/sigs.k8s.io/structured-merge-diff/v4/merge/update.go#L209-L211
In our use case, the 2nd field manager applies the same configuration as the 1st, so the merged object is indeed the same as the current live object, so a `nil` is indeed returned to signal that.

BTW I believe there was a typo in the go doc comments of the `Manager` interface’s `Apply` method. Latest push fixed the typo.

Fully agree that we need unit tests here. Will work on that.

Reply:
I think the logic in this PR is correct, but needs commenting to explain.

I'd suggest:
```go
	if object != nil {
		// non-nil object means the apply operation modified the object, so update the manager's timestamp
		managed.Times()[fieldManager] = &metav1.Time{Time: time.Now().UTC()}
	} else {
		// nil object means the apply operation did not modify the input object
		// clone the input object and return it without managed fields
		object = liveObj.DeepCopyObject()
		RemoveObjectManagedFields(object)

		if _, managerInFields := managed.Fields()[fieldManager]; managerInFields {
			if _, managerInTimes := managed.Times()[fieldManager]; !managerInTimes {
				// if the manager owns fields, ensure it has an associated time.
				// a no-op apply can add co-ownership of existing field values, so record the time that occurred.
				managed.Times()[fieldManager] = &metav1.Time{Time: time.Now().UTC()}
			}
		}
}
```